prov/gni/src/gnix_ep.c

```diff
@@ -576,6 +588,8 @@ int gnix_ep_open(struct fid_domain *domain, struct fi_info *info,
 		}
 	}
 
-	if (nic != NULL) free(nic);
+	if (nic != NULL)
+		free(nic);
+	pthread_mutex_unlock(&gnix_nic_list_lock);
```

Comment:
This is rather far away from the lock. Would it possibly improve performance to acquire the lock only around when performing operations? I'm not sure whether or not it'd make a difference, but just a consideration...